### DISCUSSION TITLE
Exception instead of NaN/Infinity

### DISCUSSION THREAD

**Discussion by osexpert** [dotnet/csharplang#2198] [Language Ideas]
Silent NaN/Inf must be one of the most stupid constructs ever?

Suggestion: on project level make it possible to enable strict FP that will throw instead of producing NaN/Inf. In places where you want to allow NaN/Inf instead of exception (when strict FP is enabled) I suggest following:

```
unsafefp
{
   // will produce NaN/Inf instead of exception
}
```

Implementation: on some arch it can be setting a FP flag in cpu and the cpu will fault. On other cpus it can be manual NaN/Inf checks after every FP operation. This will be done by individual runtimes. The unsafefp block translate to two compiler generated calls to the runtime that turn strictfp off and on (in try/finally fashion). If strict FP project setting is off (default), no such compiler generated calls are generated (100% opt-in functionality). Will need a new runtime method eg. SetStrictFP(bool enable).

Easy peasy:-)

Keywords:
NaN considered harmful
Infinity considered harmful

**Comment by CyrusNajmabadi** [comment]
> Silent NaN/Inf must be one of the most stupid constructs ever?

I don't know what would lead to that conclusion.  NaN/Inf are how FP makes sense mapping an infinite space to a fixed-size representation.  

They are all well defined in IEEE754, and are properly understood and propagated through all calculations as would be expected.

**Comment by CyrusNajmabadi** [comment]
> Suggestion: on project level

Project level suggestions are not part of the language (which has no concept of 'projects').  If you want a compiler option, you can suggest it at dotnet/roslyn.

**Comment by MkazemAkhgary** [comment]
If you are concerned you should do the checks where you think division by 0 happens or anything unexpected.

NaN is easy to find. it's like a virus and spreads every where, so you don't have to check numbers for being NaN after every floating point operation. (for example you can put checks at end of the loop)

notice that these checks will slow down your program if you are dealing with many floating point operations. I think that is one reason to why NaN/Inf is silent. much more important reason is that NaN and Inf are both existing concepts in math.

At the end, you have the knowledge of where your calculations can actually become NaN/Inf. it's your responsibility to strictly put checks where ever required.

I have made a small struct for you that you can use for debug purposes. this construct is made so that your code would be very similar to double usage (so you can quickly make a release mode)

```c#
public struct FiniteDouble
{
    double val;

    public FiniteDouble(double d)
    {
        if (double.IsNaN(d)) throw new ArgumentOutOfRangeException("double is Not a Number.");
        if (double.IsInfinity(d)) throw new ArgumentOutOfRangeException("double is Infinite.");

        val = d;
    }

    // implicit cast is safe, exception is thrown if number is not finite.
    public static implicit operator FiniteDouble(double d)
    {
        return new FiniteDouble(d);
    }

    // explicit cast because we dont want to accidentally get out of finite range.
    public static explicit operator double(FiniteDouble d)
    {
        return d.val;
    }

    public static FiniteDouble operator +(FiniteDouble x, FiniteDouble y)
    {
        return new FiniteDouble(x.val + y.val);
    }

    public static FiniteDouble operator -(FiniteDouble x, FiniteDouble y)
    {
        return new FiniteDouble(x.val - y.val);
    }

    public static FiniteDouble operator *(FiniteDouble x, FiniteDouble y)
    {
        return new FiniteDouble(x.val * y.val);
    }

    public static FiniteDouble operator /(FiniteDouble x, FiniteDouble y)
    {
        return new FiniteDouble(x.val / y.val);
    }

    public static FiniteDouble Log(FiniteDouble x)
    {
        return Math.Log((double)x);
    }

    public static FiniteDouble Log(FiniteDouble x, FiniteDouble newBase)
    {
        return Math.Log((double)x, (double) newBase);
    }

    // etc
}
```

Here is an example

```c#
FiniteDouble x = 435d; // just change FiniteDouble to double
x *= 10000;
x /= 0;
x = x + 3453;
```

Once again, doing checks after every operation can be costly depending on your usage. structs are value types and for this particular struct it has no overhead by it self (its just a double), the problem is those checks in constructor.

**Comment by orthoxerox** [comment]
@CyrusNajmabadi technically, IEEE 754 allows the implementation to treat FP exceptions as errors provided it's not the default mode. It could work similar to `checked/unchecked`.

**Comment by osexpert** [comment]
> 
> 
> > Suggestion: on project level
> 
> Project level suggestions are not part of the language (which has no concept of 'projects'). If you want a compiler option, you can suggest it at dotnet/roslyn.

unsafefp keyword is part of language. the methods to turn it off/on will be part of runtime. plus the project setting. it is pretty similar to checked/unchecked as @orthoxerox said. but it may not belong here.

**Comment by osexpert** [comment]
> 
> 
> @CyrusNajmabadi technically, IEEE 754 allows the implementation to treat FP exceptions as errors provided it's not the default mode. It could work similar to `checked/unchecked`.

Yes, for arch\cpu that support it, the runtime would utilize that. For others the runtime must do "manual" NaN/Inf checks.

**Comment by osexpert** [comment]
> 
> 
> If you are concerned you should do the checks where you think division by 0 happens or anything unexpected.
> 
That will be some million checks in our code:-)
> NaN is easy to find. it's like a virus and spreads every where, so you don't have to check numbers for being NaN after every floating point operation. (for example you can put checks at end of the loop)
> 
My last NaN encounter was an integrator that went into infinite loop because of invalid input. I have many stories like that.
> notice that these checks will slow down your program if you are dealing with many floating point operations. I think that is one reason to why NaN/Inf is silent. much more important reason is that NaN and Inf are both existing concepts in math

Yes perf can be a problem, but its possibly to use unsafefp at hotspots. And some cpu may have option to enable FP unit to fault it.
> 
> At the end, you have the knowledge of where your calculations can actually become NaN/Inf. it's your responsibility to strictly put checks where ever required.
>
Yes I know, but that will be a lot of checks. In practice these checks are just sometimes added and mostly  turn into bugs. 
> I have made a small struct for you that you can use for debug purposes. this construct is made so that your code would be very similar to double usage (so you can quickly make a release mode)
> 
Nice. Thanks!
> Once again, doing checks after every operation can be costly depending on your usage. structs are value types and for this particular struct it has no overhead by it self (its just a double), the problem is those checks in constructor.

**Comment by CyrusNajmabadi** [comment]
> My last NaN encounter was an integrator that went into infinite loop because of invalid input. I have many stories like that.

A algorithm that doesn't properly handle the spec'ed behavior of the domain it operates on will def have problems.  People have had similar problems not understanding how to properly do FP comparisons and whatnot.  

My advice would be to study up *a lot* on IEEE754 and all it's behaviors.  Then write your code according to that.

**Comment by CyrusNajmabadi** [comment]
> unsafefp keyword is part of language.

I don't know what that is referring to.

**Comment by CyrusNajmabadi** [comment]
> Yes I know, but that will be a lot of checks. In practice these checks are just sometimes added and mostly turn into bugs.

If you are actually concerned, take the above approach, but only do the checks in debug builds.  Test thoroughly, and use the checks to find issue.  Stamp them out.  Then rebuild in 'release' mode where it doesn't actually do the check.

in a similar, vein, you can just use this type, but have it enable the underlying hardware checks if available, or else it will do the checks itself.

**Comment by osexpert** [comment]
> 
> 
> > unsafefp keyword is part of language.
> 
> I don't know what that is referring to.

As I wrote in 1st post, unsafefp is the suggested new keyword\language feature.

**Comment by CyrusNajmabadi** [comment]
Can you place that code in an ``` block?

**Comment by CyrusNajmabadi** [comment]
> Implementation: on some arch it can be setting a FP flag in cpu, on others it can be manual NaN/Inf checks after every FP operation.

This seems... undesirable.  If you want this behavior, why do you need language support?  Why not just create a library that you call into to 'flip/enable' this behavior?  That way you can decide if the impact here is appropriate for your needs.

**Comment by osexpert** [comment]
> 
> 
> Can you place that code in an ``` block?

done

**Comment by osexpert** [comment]
> 
> 
> > Implementation: on some arch it can be setting a FP flag in cpu, on others it can be manual NaN/Inf checks after every FP operation.
> 
> This seems... undesirable. If you want this behavior, why do you need language support? Why not just create a library that you call into to 'flip/enable' this behavior? That way you can decide if the impact here is appropriate for your needs.

Not sure what you mean. Where can I flip this behavior? Its not part of the runtime afaik.
But I think I understand: most of the suggested function belong in runtime, it is only the unsafefp "sugar" (that could be simulated with try/finally) that is the language feature.

**Comment by DavidArno** [comment]
It's worth pointing out that your suggestion of,
```cs
unsafefp
{
   // will produce NaN/Inf instead of exception
}
```

is a definite non-starter. You are changing the default behaviour of FP, which would cause code that currently runs to now fail with exceptions. If this idea were considered, it would have to be an opt-in with the current behaviour being the default.

**Comment by yaakov-h** [comment]
@DavidArno They've already suggested a project-level switch (which I assume would end up as a command-line argument to `csc.exe`) so assuming that the switch is opt-in at the project level, this proposal is no more of a non-starter / change-of-defaults than `checked`/`unchecked` is today.

**Comment by YairHalberstadt** [comment]
It should essentially have `checked`, `unchecked` semantics.

**Comment by yaakov-h** [comment]
Essentially I think it's just a floating-point version of `checked`/`unchecked`.

It's a shame that adding it to `checked` would be a breaking change, but as its own keyword you would need to apply both `checked`/`unchecked` and the floating-point equivalents each time you want to turn on or off safety.

Having two sets of safety mechanisms seems like a bit much.

**Comment by tannergooding** [comment]
This should probably be an API proposal against CoreFX requesting that the framework/runtime expose a mechanism to allow IEEE compliant exceptions. Not only is this non C# specific, but it requires runtime changes to enable/support. The proposal could start off as a discussion initially, but it eventually needs to be a formal API proposal (the API review process and a template to follow for the issue can be found here: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md#steps).

The runtime spec indicates:
> While the IEC 60559:1989 standard also allows for exceptions to be thrown under unusual conditions (such as overflow and invalid operand), the CLI does not generate these exceptions.
Instead, the CLI uses the NaN, +infinity, and –infinity return values and provides the instruction ckfinite to allow users to generate an exception if a result is NaN, +infinity, or –infinity.

I have been working on updating the various code in CoreFX/CoreCLR to be IEEE 754:2008 compliant. The issue tracking this support is here: https://github.com/dotnet/corefx/issues/31901. However, `exception` support was considered "out-of-scope" at this time since it does require significant design/support/etc and since the majority of customers end up not wanting it and instead (much like with `checked` vs `unchecked`) preferring speed over safety checks. This lead us to determine that the cost-to-benefit ratio was not worthwhile and we should come back if and when we had customers ask for the feature.

It is worth noting that, if the language (or framework/runtime) were to expose some mechanism to allow toggling IEEE compliant exceptions, this proposal is not currently sufficient as the IEEE 754:2008 standard defines a number of exceptions each of which can be toggled on/off independently from the others. A user wanting the current behavior (no exceptions) would be just as valid as a user wanting exceptions for Overflow/NaN (this proposal), users wanting exceptions for Underflow (not listed as part of this proposal, but also generally applicable), users wanting exceptions only for NaN, users wanting signalling for "inexact" results, etc. The total exceptions defined are: `Invalid Operation`, `Division By Zero`, `Overflow`, `Underflow`, and `Inexact`.

It is also worth noting that, for IEEE compliant exceptions; `NaN` and `Infinity` are not themselves invalid. Each operation has its own behavior that may deviate from your expected result. For example: most say `NaN` is valid and will only throw for `SNaN`. Most also consider `Infinity` to be a valid input and only consider overflow when all non-finite inputs produce an infinitely precise result (which is not `Infinity`), which must then round to `Infinity` to be represented in the destination format (`Log2(0)` for example is exactly `Infinity`, and would not be considered `Overflow`; the same with `double.PositiveInfinity + 2.0`).

**Comment by YairHalberstadt** [comment]
The details sound horribly complex, and would probably cause a huge amount of confusion along the lines of "why are/nt I getting an exception here".

Whilst initially I thought that this would be a reasonable idea if it could be done easily, now I think that the extra complexity of this feature are not worth it.

**Comment by tannergooding** [comment]
> The details sound horribly complex, and would probably cause a huge amount of confusion along the lines of "why are/nt I getting an exception here".

Yes. IEEE floating-point is already fairly confusing by itself. For example, plenty of people don't understand that `double d = literal` does not give you exactly `literal`, but instead `literal` rounded to the nearest representable result.

Once you start adding in exceptions, rounding modes, etc. it just gets more confusing.

**Comment by bbarry** [comment]
Supposing the language were able to enable IEEE compliant exception behavior, what might that look like?

Presumably the semantics would be something like a using statement at the block level:

```csharp
using (FloatingPointCompliance.Enable(invalidOperation: true, divideByZero: true, overflow: true, underflow: true, inexact: true)) {
    ...
}
```

or maybe an attribute at the method level?

That doesn't seem to require a language syntax change outside of mentioning a special named type that the CLR should react to. Perhaps one would be nice to sugar over such an implementation?

**Comment by osexpert** [comment]
So it may seem the FPU doesn't have an option for fautling at !IsFinite (per @tannergooding comment). In case the only remaining alternative is IsFinite checks generated by the runtime after every FP operation that can manifest them (it may not be than many?). So it is doable off course. And it will hurt perf. It's not as generic as you want it to be (adhering to IEEE defined exceptions). But I can understand you want a solution that not only fixes my problem:-)

Thank you for the consideration.

**Comment by Thaina** [comment]
> NaN considered harmful

probably

> Infinity considered harmful

Totally wrong

**Comment by GSPP** [comment]
I am very sympathetic to the idea of disallowing "unusual" floating point values. I have wanted this multiple times. Most of the time any such value at any point indicates a bug.

I'd argue that *most* code wants to use only finite floats. Certainly any kind of business logic style code (don't say you should use decimal, I'm not talking about money). Math-heavy code likely wants to allow non-finite values for performance.

You can of course check the end result but it can amount to a lot of checking. It's easy to forget and the check does not flag where the bug occurred (it might occur in an intermediate step). If you only have the exception from the production logs you'd like the exception to occur at the line where the first invalid value was created.

`checked` arithmetic is useful for integers. At the CLR level it's a per-operation check. Could be the same for floats. Certainly, there should *not* be any process-global setting. That would alter the behavior of other code in the process.

The IEEE standard defines all kind of behavior and different exceptions and modes. It is not necessary to expose all of this. It could be a single type of exception thrown for any kind of non-finite value.

I like the idea of the `FiniteDouble` proposed above. It's easy to implement. Any runtime changes would likely not carry their weight.

**Reply by AartBluestoke** [reply]
"Math-heavy code likely wants to allow non-finite values for performance."
the 'for performance' almost surely rules out regular checks per operation for "unusual" values without direct hardware support.

**Comment by kellypleahy** [comment]
Forgive me for being late to the party, but I found this when looking for a way to use/enable SNaNs in .NET.

I think this request is pretty poorly worded so I thought I might try to help with the understanding and maybe resurrect it under these terms.

Assumptions:
1. Many programs that do mathematical calculations would benefit from the use of SNaNs supported by platforms / architectures that .NET runs on.  Specifically, x86/x64 definitely support this functionality.
2. There would be no adverse performance impact to applications running on platforms with native IEEE SNaN exception support
3. For applications running on platforms without this support, it seems like a graceful fallback is probably not necessary, and it would be fair to make this the application's problem (to only use this functionality where it exists).  This is similar to using things like P/Invoke against windows APIs in my mind.

Concerns:
1. Applications need a way to turn on/off this behavior easily during program execution (i.e. it's not like large address aware or similar type flags that are EXE-level)
2. The default should be what is happening now so existing apps are not affected
3. The method of implementation chosen should probably be limited in scope as much as possible, as there's a lot of moving parts in the ecosystem

My opinion / recommendation:
1. Consider doing this with library support as much as possible.  For example, a using block with a "ConfigureFPU" object or something might be a good way to do this, as mentioned here: https://github.com/dotnet/csharplang/discussions/2198#discussioncomment-107616 (above)
2. The exception thrown by the FPU will need to be handled at a low-level by the application and marshalled to some sort of .NET exception that provides the information necessary for the user (application programmer) to understand what happened and where.  I don't think there are great ways to do this for "intra-expression" errors that give you anything other than a "per-statement" exception (i.e. you won't know which operation in an expression like a/b + c/d caused the "divide by zero" exception) but this isn't special to these expressions - I think this is true for integer expressions or complex expressions involving functions that throw exceptions already so it's reasonable to have this behavior here too.  There is probably some decoding of some low-level registers / flags in the FPU to determine how to map the exception to something meaningful (my guess is this will come across as some sort of SEH exception from windows, and I'm not sure how it manifests on linux implementations).

My reason for wanting this:
Similar to "checked null" in modern reference types in C#, I think it would be immensely valuable for people writing complex scientific and mathematical code (much like what I write on a daily basis) to be able to turn on checked NaNs in places where this would help identify immediately where things got "bad".  In fact, working with NaN is much HARDER than working with nulls, as null exceptions manifest themselves as soon as the code tries to dereference the pointer.  This is easier to find (while not perfect) than the case of a NaN, as sometimes it's not immediately clear what is the "first" place the NaN came into the calculation.  Being able to turn on checked NaN (i.e. turn off "unmasking" of signalling NaN in FPU terms) would allow the developer to quickly find exactly where in the code the NaN first appeared and diagnose problems much more quickly.  Otherwise one must resort to reading the code and reading the logs and trying to reason about where the problem could have occurred.

Thanks and sorry for the huge post!

**Reply by tannergooding** [reply]
> There would be no adverse performance impact to applications running on platforms with native IEEE SNaN exception support

I would guess this is essentially a non-starter as exceptions in .NET tend to be expensive. I imagine floating-point exceptions would work similarly to `NullReferenceException` in that the underlying hardware fault would be converted into a .NET exception and so they would have similar performance characteristics (cheap to check, expensive if thrown).

Additionally, while hardware tends to operate fast with floating-point exceptions enabled or disabled (provided no exception is thrown), actually transitioning the floating-point control state between "enabled" and "disabled" can be quite expensive.

> Applications need a way to turn on/off this behavior easily during program execution

I think an even more key part to this is scoping and maintaining compatibility with existing .NET code.

That is, it isn't necessarily as simple as just enabling it for a region of code and then disabling it when done. There are libraries/apps that rely on `NaN` already today (like WPF) and where code such as the following *must* maintain the `FpExceptions=Disabled` state in `CallToNonFpExceptionFunction` or by enabling floating-point exceptions, you risk breaking arbitrary .NET Code:
```
EnableFpExceptions();
CallToNonFpExceptionFunction();
DisableFpExceptions();
```

A similar issue exists with the proposal to allow "fast math" floating-point operations or to support specified rounding modes. You effectively must maintain back compat with any existing .NET code (we'll call this `floating-point mode: unaware`).
You then can opt specific methods into new control modes, but you must go back to the "default" mode at call boundaries (assuming the modes don't match or can't be known to match) or during inlining.

**Reply by kellypleahy** [reply]
@tannergooding Thanks for the extremely well-thought-out reply.  I don't think I agree with your first point, as the exceptions would be in cases where you want to trigger a "stop" anyway, but your other points are excellent and I agree with them completely.

Those are specific aspects of the problem I had not previously considered, and to work around them would require much more intelligent language support than the library approach I had considered.

Thanks for the response.

As an additional note to your last comment, I would be happy personally if this applied to call boundaries, but can understand the performance impact that would impart, and the complexity of it as well, considering you would need some sort of language / runtime enforcement and much more intelligent exception handling.

**Comment by prjm** [comment]
Just a small comment from my side: In Delphi, you can change the behaviour of floating point exceptions via the runtime in a non-threadsafe manner. This leads to strange bugs like this one: [Floating point exception when embedding IE](https://stackoverflow.com/questions/19187479/masking-floating-point-exceptions-with-set8087cw-setmxcsr-and-twebbrowser). This is certainly not a success story, so if something changes here in C#, it has  to be designed carefully.
